File: programmers/c++/hash/Programmers_best_album.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>

using namespace std;

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    int check_genres[10000] = { 0 };
    vector<string>kind;
    map<int, tuple<string, int>> genres_map;
    map<string, int> views;
    for (int i = 0; i < genres.size(); i++) {
        if (views.find(genres[i]) == views.end()){
            views.insert(make_pair(genres[i], plays[i]));
            kind.push_back(genres[i]);
        }else
            views[genres[i]] += plays[i];
        genres_map[i] = make_tuple(genres[i],plays[i]);
    }

    while (kind.size()) {
        string max_kind = kind[0];
        for (auto &kind : kind) {
            if (views[max_kind] < views[kind])
                max_kind = kind;
        }
        int two_count = 0;
        while (two_count < 2) {
            int i = -1;
            for (int t = 0; t < genres_map.size(); t++) {
                if (get<0>(genres_map[t]) == max_kind && !check_genres[t]) {
                    if (i == -1)
                        i = t;
                    else if (get<1>(genres_map[i]) < get<1>(genres_map[t]))
                        i = t;      
                }
            }
            if (i != -1) {
                answer.push_back(i);
                check_genres[i] = 1;
                two_count++;
            }else {
                break;
            }
           
        }
        for (int i = 0; i < kind.size(); i++) {
            if (kind[i] == max_kind) {
                kind.erase(kind.begin() + i);
                break;
            }
        }
    }

    return answer;
}

File: programmers/python/hash/Programmers_bestalbum.py
def solution(genres, plays):
    answer = []
    onecchk=[0 for i in range(len(genres))]
    Dgenres = {}
    for g,p in zip(genres, plays):
        if g in Dgenres:
            Dgenres[g].append(p)
        else:
            Dgenres[g] = [p]
    for g in genres:
        Dgenres[g] = sorted(Dgenres[g], reverse=True)
    Dgenres = sorted(Dgenres.items(), reverse=True, key=lambda x:sum(x[1]))
    for d in Dgenres:
        for nd in d[1][:2]:
            for a in range(len(genres)):
                if d[0] == genres[a] and nd == plays[a] and onecchk[a] == 0:
                  answer.append(a)
                  onecchk[a] = 1   
                  break     
    return answer
